### PR TITLE
test_username_api.py: Remove duplicated logic

### PR DESCRIPTION
**Pull request by seeeturtle**
This will remove duplicated logic
and put into a function for each test.

Closes https://github.com/manu-chroma/username-availability-checker/issues/36

### CODE REVIEW

File: tests/test_username_api.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import logging
 import username_api
 import os.path
+from parameterized import parameterized
```
Comment (manu-chroma) [inline tests/test_username_api.py]: separate out imports into: 
- standard lib imports
- pip installed imports
- custom(own) modules import

File: tests/test_username_api.py [hunk 2]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
-    del actual['profile']
-
-    message = None
-    if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
-
-      user = generate_random_username(25)
-      expected = get_expected_response('facebook', user, 404)
-      actual = get_response(self.app, 'facebook', user)
+  @parameterized.expand(load_test_cases('available_username'), testcase_func_name=custom_name_func)
```
Comment (manu-chroma) [inline tests/test_username_api.py]: split like this: https://github.com/manu-chroma/username-availability-checker/blob/master/tests/test_get_avatar.py#L41

File: tests/test_username_api.py [hunk 3]
```diff
@@ -351,94 +108,17 @@ def test_available_deviantart_username(self, user):
...
+    assert_response(self.app, website, user, 200)
 
-  def test_deviantart_format_checking(self):
-    resp = self.app.get('/check/deviantart/{}'.format(invalid_username))
+  @parameterized.expand(load_test_cases('format_checking'), testcase_func_name=custom_name_func)
+  def test_format_checking(self, website, username):
+    resp = self.app.get('/check/{w}/{u}'.format(w=website, u=username))
     json_resp = json.loads(resp.get_data().decode())
+    u = self.websites['urls'].get(website, 'https://{w}.com/{u}')
+    url = u.format(w=website, u=username)
```
Comment (manu-chroma) [inline tests/test_username_api.py]: why this?

File: tests/test_username_api.py [hunk 4]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
+        usernames = invalid_username_facebook
+      else:
+        usernames = [invalid_username]
+    res.extend((website, user) for user in usernames)
+  
+  return res
+
+
+def custom_name_func(testcase_func, param_num, param):
+  logging.getLogger().warn(param.args)
```
Comment (manu-chroma) [inline tests/test_username_api.py]: don't think we need this

File: tests/test_username_api.py [hunk 5]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
-      user = generate_random_username(25)
-      expected = get_expected_response('facebook', user, 404)
-      actual = get_response(self.app, 'facebook', user)
+  @parameterized.expand(load_test_cases('available_username'), testcase_func_name=custom_name_func)
+  def test_available_name(self, website, user):
+    expected = get_expected_response(website, user, 404)
+    actual = get_response(self.app, website, user)
 
+    if website == 'facebook':
       del actual['profile']
```
Comment (manu-chroma) [inline tests/test_username_api.py]: don't delete the comment attached to this. pls re-add it

File: tests/test_username_api.py [hunk 6]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['facebook']['available_usernames'])
-  def test_available_facebook_username(self, user):
-    expected = get_expected_response('facebook', user, 404)
-    actual = get_response(self.app, 'facebook', user)
-
-    # this is special to facebook checking,
-    # and 'profile' field is only useful for
-    # frontend purposes. Skipping here.
```
Comment (manu-chroma) [inline tests/test_username_api.py]: don't delete this comment

File: tests/test_username_api.py [hunk 7]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
-
-    message = None
-    if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
-
-      user = generate_random_username(25)
-      expected = get_expected_response('facebook', user, 404)
-      actual = get_response(self.app, 'facebook', user)
+  @parameterized.expand(load_test_cases('available_username'), testcase_func_name=custom_name_func)
+  def test_available_name(self, website, user):
```
Comment (manu-chroma) [inline tests/test_username_api.py]: ``test_available_username``

File: tests/test_username_api.py [hunk 8]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
 
+def load_test_cases(type):
+  res = []
+  for website in data:
+    if type == 'available_username':
+      usernames = data[website]['available_usernames']
+    if type == 'taken_username':
+      usernames = data[website]['taken_usernames']
+    if type == 'format_checking':
+      if website == 'facebook':
```
Comment (manu-chroma) [inline tests/test_username_api.py]: you can do an append to ``usernames`` list ``if website == 'facebook'`` and bring the facebook condition down.

File: tests/test_username_api.py [hunk 9]
```diff
@@ -351,94 +108,17 @@ def test_available_deviantart_username(self, user):
...
-      'possible': False,
-      'url': 'https://www.openhub.net/accounts/{}'.format(invalid_username)
-    } == json_resp
+  @parameterized.expand(load_test_cases('taken_username'), testcase_func_name=custom_name_func)
+  def test_taken_username(self, website, user):
+    assert_response(self.app, website, user, 200)
 
-  def test_deviantart_format_checking(self):
-    resp = self.app.get('/check/deviantart/{}'.format(invalid_username))
+  @parameterized.expand(load_test_cases('format_checking'), testcase_func_name=custom_name_func)
```
Comment (manu-chroma) [inline tests/test_username_api.py]: https://travis-ci.org/manu-chroma/username-availability-checker/builds/320656667#L641

is ``re.sub`` happening in the actual data or just when creating the function name?
pls confirm this

File: tests/test_username_api.py (hunk 3 above)
Comment (nalinbhardwaj) [reply tests/test_username_api.py]: You probably want to use [username_api.get_profile_url](https://github.com/manu-chroma/username-availability-checker/blob/master/username_api.py#L17)

File: tests/test_username_api.py [hunk 10]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
+  return '%s_%s' % (
+    testcase_func.__name__,
+    parameterized.to_safe_name('_'.join(str(x) for x in param.args)),
+  )
+
 class TestUsernameApi(object):
   @classmethod
   def setup_class(self):
     self.app = username_api.app.test_client()
     self.data = data
```
Comment (nalinbhardwaj) [inline tests/test_username_api.py]: seems unused.

File: tests/test_username_api.py [hunk 11]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
   }
 
+def load_test_cases(type):
+  res = []
+  for website in data:
+    if type == 'available_username':
+      usernames = data[website]['available_usernames']
+    if type == 'taken_username':
+      usernames = data[website]['taken_usernames']
+    if type == 'format_checking':
```
Comment (nalinbhardwaj) [inline tests/test_username_api.py]: Better to break this into it's separate function.

File: tests/test_username_api.py [hunk 12]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
-      user = generate_random_username(30, first_char_letter=True)
-      expected = get_expected_response('openhub', user, 404)
-      actual = get_response(self.app, 'openhub', user)
+      username_pattern = self.websites['username_patterns'][website]
+      random_length = random.randint(username_pattern['min_length'], username_pattern['max_length'])  
 
-      if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
+      if website == 'openhub':
+        user = generate_random_username(random_length, first_char_letter=True)
```
Comment (nalinbhardwaj) [inline tests/test_username_api.py]: I think it's a generate_random_username problem, not a testing problem. Better to modify the function to take the website as parameter, and process accordingly(not do this here)

File: tests/test_username_api.py [hunk 13]
```diff
@@ -40,306 +41,62 @@ def get_expected_response(website, user, status):
...
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['openhub']['available_usernames'])
-  def test_available_openhub_username(self, user):
-    expected = get_expected_response('openhub', user, 404)
-    actual = get_response(self.app, 'openhub', user)
-
     message = None
     if expected != actual:
       message = 'The provided available username ({}) returned 200'.format(user)
```
Comment (nalinbhardwaj) [inline tests/test_username_api.py]: Is there any point in having these messages now? We can already see the used username with `parameterized`, and know that it's failing...

File: tests/test_username_api.py (hunk 3 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: thanks :)

File: tests/test_username_api.py (hunk 4 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: sorry, forgot to remove.

File: tests/test_username_api.py (hunk 6 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: sure

File: tests/test_username_api.py (hunk 7 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: what do you mean?

File: tests/test_username_api.py (hunk 8 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: hmm... do you mean to filter `facebook` at below?

File: tests/test_username_api.py (hunk 9 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: https://docs.python.org/3/library/re.html?highlight=re%20sub#re.sub according to python doc re.sub just "returns" not modify.

File: tests/test_username_api.py (hunk 11 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: umm... breaking into `format_checking` and `taken/available_username`?

File: tests/test_username_api.py (hunk 12 above)
Comment (nalinbhardwaj) [reply tests/test_username_api.py]: But this is not a very scalable thing, right? As in, we can't expect to have if elif's for tons of websites in the future, it would just make the function clunky. Having it in it's separate function to satisfy each website's own quirks is better(since they will only grow)

File: tests/test_username_api.py (hunk 11 above)
Comment (nalinbhardwaj) [reply tests/test_username_api.py]: Indeed, `load_format_checking_cases` and `load_general_test_cases`(or something more descriptive), would be clearer, since format checking is most independent of taken/not taken usernames

File: tests/test_username_api.py (hunk 12 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: But passing website name will be same since it should check what options are needed for passed website.

Comment (seeeturtle) [reply tests/test_username_api.py]: if elif statements are unavoidable.

File: tests/test_username_api.py (hunk 11 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: alright

File: tests/test_username_api.py (hunk 7 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: ah, I got it.

File: tests/test_username_api.py (hunk 12 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: anyway, I think your way will look better.

File: tests/test_username_api.py [hunk 14]
```diff
@@ -1,24 +1,29 @@
-import unittest
 import json
-import yaml
-import string
+import logging
 import random
+import string
+import os.path
```
Comment (jayvdb) [inline tests/test_username_api.py]: as you are moving imports around, please move them into alpha order per PEP8

File: tests/test_username_api.py [hunk 15]
```diff
@@ -1,24 +1,29 @@
...
-def generate_random_username(length, first_char_letter=False):
-  return ''.join(generate_random_character(length, first_char_letter))
+def generate_random_username(website, length):
+  if website == 'openhub':
+    first_char_letter = True
+  return ''.join(generate_random_character(length,
+    first_char_letter=first_char_letter))
 
-def generate_random_character(length, first_char_letter):
+def generate_random_character(length, **kwargs):
```
Comment (jayvdb) [inline tests/test_username_api.py]: there is no benefit in using kwargs here.

and fiddling with this random generator is not related to your issue.

File: tests/test_username_api.py [hunk 16]
```diff
@@ -40,405 +45,87 @@ def get_expected_response(website, user, status):
...
+    if status == 'taken_username':
+      usernames = data[website]['taken_usernames']
+    res.extend((website, user) for user in usernames)
+
+  return res
+
+def load_format_checking_cases():
+  res = []
+  for website in data:
+    if website == 'facebook':
```
Comment (jayvdb) [inline tests/test_username_api.py]: data belongs in the yaml, not here in code.

File: tests/test_username_api.py [hunk 17]
```diff
@@ -40,405 +45,87 @@ def get_expected_response(website, user, status):
...
-      logging.getLogger().warn(message)
-
-    assert expected == actual
-
-  @pytest.mark.parametrize('user', data['deviantart']['available_usernames'])
-  def test_available_deviantart_username(self, user):
-    expected = get_expected_response('deviantart', user, 404)
-    actual = get_response(self.app, 'deviantart', user)
+      username_pattern = self.websites['username_patterns'][website]
+      random_length = random.randint(username_pattern['min_length'],
```
Comment (jayvdb) [inline tests/test_username_api.py]: improving the generation of usernames is not part of this issue.
please remove it.
focus your commits.
We can create separate tasks for things like this.

File: tests/test_username_api.py (hunk 14 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: sure

File: tests/test_username_api.py (hunk 15 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: I used kwargs since we can add more option later. It's quite possible.

File: tests/test_username_api.py (hunk 16 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: umm... what do you mean?

File: tests/test_username_api.py (hunk 17 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: @jayvdb can you create a separate task?

Comment (seeeturtle) [reply tests/test_username_api.py]: then I'll set length that just fit with every website.

File: tests/test_username_api.py (hunk 14 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: done

File: tests/test_username_api.py (hunk 15 above)
Comment (jayvdb) [reply tests/test_username_api.py]: No, thank you.  This is unrelated to this change.
This isnt even the right way to do it.
I've created https://github.com/manu-chroma/username-availability-checker/issues/47 for that.

File: tests/test_username_api.py (hunk 16 above)
Comment (jayvdb) [reply tests/test_username_api.py]: put these usernames in https://github.com/manu-chroma/username-availability-checker/blob/master/tests/test_data.yml

File: tests/test_username_api.py (hunk 15 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: got it.

File: tests/test_username_api.py (hunk 16 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: got it

Comment (seeeturtle) [reply tests/test_username_api.py]: why don't we add new key, `invalid_usernames`? then we can remove this if statement.

File: tests/test_username_api.py [hunk 18]
```diff
@@ -40,405 +46,83 @@ def get_expected_response(website, user, status):
...
 
-      user = generate_random_username(15)
-      expected = get_expected_response('deviantart', user, 404)
-      actual = get_response(self.app, 'deviantart', user)
+      if website == 'facebook':
+        del actual['profile']
 
       if expected != actual:
-        message += ' and the random username ({}) returned 200'.format(user)
+        message = 'the random username ({}) returned 200'.format(user)
```
Comment (jayvdb) [inline tests/test_username_api.py]: why are you changing these messages?

Comment (seeeturtle) [reply tests/test_username_api.py]: refer to @nalinbhardwaj's review
> Is there any point in having these messages now? We can already see the used username with parameterized, and know that it's failing...

Comment (jayvdb) [reply tests/test_username_api.py]: The function name cant be seen except in verbose mode.

File: tests/test_username_api.py [hunk 19]
```diff
@@ -40,405 +46,83 @@ def get_expected_response(website, user, status):
...
+      username_length = 6
 
-    message = None
-    if expected != actual:
-      message = 'The provided available username ({}) returned 200'.format(user)
+      user = generate_random_username(website, username_length)
+      expected = get_expected_response(website, user, 404)
+      actual = get_response(self.app, website, user)
 
-      user = generate_random_username(15)
```
Comment (jayvdb) [inline tests/test_username_api.py]: why did you change this from 15 to 6?

File: tests/test_username_api.py (hunk 13 above)
Comment (jayvdb) [reply tests/test_username_api.py]: Late reply.

Firstly, very interesting review comment.

Personally, I would not have recommended this change because it means the test result message isnt self-contained, and doesnt really make sense without the verbose format, or another plugin which uses the test function name in the output.

I think we could also uses a test function docstring generator to make this change compliant with [`pytest-spec`](https://pypi.python.org/pypi/pytest-spec) , but ensuring compatibility with that would be optional for this task.

I am not going to reject the task based on this, as the goal has been reached and this change was requested, reasonable and done correctly.  It definitely isnt 'wrong'.  I suspect this is a 'matter of taste' and I will defer to @manu-chroma on this one.

File: tests/test_username_api.py (hunk 19 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: 15 is for `test_available_deviantart_username`. 6 is just valid minimum length.

File: tests/test_username_api.py (hunk 18 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: I thought function name will be shown if test fails. isn't it?

File: tests/test_username_api.py (hunk 13 above)
Comment (seeeturtle) [reply tests/test_username_api.py]: ok

File: tests/test_username_api.py (hunk 19 above)
Comment (jayvdb) [reply tests/test_username_api.py]: 15 was also the minimum length, previously.  Some had higher, but 15 was the lowest value, and is the length of `zNRe3jx3isA8CoM` used in the test data.

Comment (seeeturtle) [reply tests/test_username_api.py]: actually, 15 is valid maximum length.

File: tests/test_username_api.py (hunk 18 above)
Comment (jayvdb) [reply tests/test_username_api.py]: It is shown, but not in the summary view, and you would be surprised how many 'developers' are scared of the pytest backtrace area (which can also be disabled; again, this concept of self-contained failure messages is relevant).

File: tests/test_username_api.py (hunk 19 above)
Comment (jayvdb) [reply tests/test_username_api.py]: The random generator always produced a username of `length` previously.  Across all invocations, previously, the minimum value for length was 15.

File: tests/test_username_api.py (hunk 13 above)
Comment (manu-chroma) [reply tests/test_username_api.py]: As discussed in Zulip chat before, ``parameterized`` sanitises the usernames before creating calling ``custom_func_name``
For eg: When testing for available username: ``manu.chroma`` for any website: 
```
tests/test_username_api.py::TestUsernameApi::test_available_username_twitter_manu_chroma test_username_api.py       
122 WARNING  The provided available username (manu.chroma) returned 200 but mLuogoEZ0qEwl68 is still available
```
There is a chance of not seeing the actual username in the function name. I think it is for the better that we keep this logging logic for now
